packages/shared/src/authz/policies/workspace/canCreateWorkspaceProject.ts

```diff
@@ -43,44 +49,45 @@ export const canCreateWorkspaceProjectPolicy: AuthPolicy<
   | InstanceType<typeof WorkspaceLimitsReachedError>
   | InstanceType<typeof ServerNoSessionError>
   | InstanceType<typeof ServerNoAccessError>
+  | ErrorsOf<typeof ensureWorkspacesEnabledFragment>
+  | ErrorsOf<typeof ensureMinimumServerRoleFragment>
+  | ErrorsOf<typeof ensureWorkspaceRoleAndSessionFragment>
 > =
   (loaders) =>
   async ({ userId, workspaceId }) => {
-    const env = await loaders.getEnv()
-    if (!userId) return err(new ServerNoSessionError())
-    if (!env.FF_WORKSPACES_MODULE_ENABLED) return err(new WorkspacesNotEnabledError())
+    const ensuredWorkspacesEnabled = await ensureWorkspacesEnabledFragment(loaders)({})
+    if (ensuredWorkspacesEnabled.isErr) return err(ensuredWorkspacesEnabled.error)
 
-    const isActiveServerUser = await hasMinimumServerRole(loaders)({
+    const ensuredServerRole = await ensureMinimumServerRoleFragment(loaders)({
       userId,
       role: Roles.Server.User
     })
-    if (!isActiveServerUser) return err(new ServerNoAccessError())
+    if (ensuredServerRole.isErr) return err(ensuredServerRole.error)
+
+    const ensuredWorkspaceAccess = await ensureWorkspaceRoleAndSessionFragment(loaders)(
+      {
+        userId: userId!,
+        workspaceId
+      }
+    )
+    if (ensuredWorkspaceAccess.isErr) {
+      return err(ensuredWorkspaceAccess.error)
+    }
 
-    const memberWithSsoSession = await maybeMemberRoleWithValidSsoSessionIfNeeded(
-      loaders
-    )({
-      userId,
-      workspaceId
-    })
     // guests cannot create projects in the workspace
-    if (memberWithSsoSession.isNothing)
+    const isNotGuest = await requireMinimumWorkspaceRole(loaders)({
```

Comment:
The previous logic of a `nothing()` actually being an error case for us was too complicated to reason about. Now `maybeMemberRoleWithValidSsoSessionIfNeeded/ ensureWorkspaceRoleAndSessionFragment` just ensures you have workspace access (any role). And the separate check of not allowing guests is its own thing.

Reply:
that, or ensureWorkspaceRoleAndSessionFragment() could have an optional `role` arg that allows you to specify the minimum role you expect